### PR TITLE
Fjord: Add FastLZ compression into L1CostFunc

### PR DESCRIPTION
**Pull request by mdehoog**
**Description**
We found that the FastLZ algorithm is a pretty good estimate for the actual results we'd see from zlib compressing the batches we write to L1 (albeit with a different `scalar` as the compression ratios aren't quite as good). See https://github.com/roberto-bayardo/compression-analysis and [this sheet](https://docs.google.com/spreadsheets/d/1vZNAtyGnMR4iGE84wV0RZ7BpohSKUCd9Lsdg0wrNk8c/edit#gid=0).

This PR introduces a `flzCompress` call into the `DataGas` part of the `L1CostFunc`. Companion `optimism` PR is here: https://github.com/ethereum-optimism/optimism/pull/9618

**Tests**

TODO

**Additional context**

The current naive L1Cost approach:
```
l1fee = ((0s in calldata) * 4 + (1s in calldata) * 16 + 188) * l1BaseFee * 0.684
```
works pretty well on average, but penalizes very compressible txs ([e.g.](https://basescan.org/tx/0x041c111f2dea51404d713ef70a31e481bd4bf637544ab2e5e5ac2c7fb1cb1614)), and undercharges incompressible txs ([e.g.](https://basescan.org/tx/0x0330a3787711c951c18df58cb7c60c2f9ac241d05bd1bafa7a9dae317a65b3ce)). This change makes the `L1CostFunc` much fairer compared to real-world L1 costs.

### CODE REVIEW

File: core/types/rollup_cost.go [hunk 1]
```diff
@@ -113,6 +121,40 @@ func NewL1CostFunc(config *params.ChainConfig, statedb StateGetter) L1CostFunc {
...
+			// function. We detect this scenario by checking if the Ecotone parameters are
+			// unset. Note here we rely on assumption that the scalar parameters are adjacent
+			// in the buffer and l1BaseFeeScalar comes first.
+			firstEcotoneBlock := l1BlobBaseFee.BitLen() == 0 &&
+				bytes.Equal(emptyScalars, l1FeeScalars[scalarSectionStart:scalarSectionStart+8])
+			if !firstEcotoneBlock {
+				l1BaseFeeScalar, l1BlobBaseFeeScalar := extractEcotoneFeeParams(l1FeeScalars)
+				return newL1CostFuncEcotone(l1BaseFee, l1BlobBaseFee, l1BaseFeeScalar, l1BlobBaseFeeScalar)
+			}
+			log.Info("using bedrock l1 cost func for first Ecotone block", "time", blockTime)
```
Comment (sebastianst) [inline core/types/rollup_cost.go]: I'm wondering if we have to be careful to also detect this edge case when Fjord and Ecotone activate at the same time.

File: core/types/rollup_cost.go [hunk 2]
```diff
@@ -71,15 +71,22 @@ var (
...
+
 	emptyScalars = make([]byte, 8)
 )
 
 // RollupCostData is a transaction structure that caches data for quickly computing the data
 // availablility costs for the transaction.
 type RollupCostData struct {
-	zeroes, ones uint64
+	zeroes, ones   uint64
+	compressedSize uint64
```
Comment (sebastianst) [inline core/types/rollup_cost.go]: Maybe we call this new field `fastlzSize` instead, to be future proof, more specific and more similar to the spec?

File: core/types/rollup_cost_test.go [hunk 3]
```diff
@@ -243,3 +283,44 @@ func TestNewL1CostFunc(t *testing.T) {
 	require.NotNil(t, fee)
 	require.Equal(t, regolithFee, fee)
 }
+
+func TestFlzCompressLen(t *testing.T) {
```
Comment (sebastianst) [inline core/types/rollup_cost_test.go:368]: Would be great if we could create a differential test (in the monorepo probably) to test this function against the Solidity implementation of FastLZ compression on some representative set of test vectors, plus some fuzz testing.

Could additionally differential-test it against another implementation as well, like [this C implementation](https://github.com/ariya/FastLZ) using cgo.

File: params/config.go
Comment (sebastianst) [inline params/config.go:1]: We also need to add override flags like in [this pr](https://github.com/ethereum-optimism/op-geth/pull/233). But we can do it in a follow up.

File: core/types/rollup_cost.go [hunk 4]
```diff
@@ -218,8 +235,44 @@ func newL1CostFuncEcotone(l1BaseFee, l1BlobBaseFee, l1BaseFeeScalar, l1BlobBaseF
...
+
+		l1CostSigned := new(big.Int).Set(l1CostIntercept)
+		l1CostSigned.Add(l1CostSigned, fastlzTerm)
+		l1CostSigned.Add(l1CostSigned, txSizeTerm)
+		l1CostSigned.Mul(l1CostSigned, l1FeeScaled)
+		l1CostSigned.Div(l1CostSigned, fjordDivisor)
+
+		if l1CostSigned.Sign() < 0 {
+			l1CostSigned.SetInt64(0)
+		}
```
Comment (sebastianst) [inline core/types/rollup_cost.go]: We realized that we need to set a sane minimum here, because it is well possible to get to zero or negative `l1CostSigned` values. When we expand the tests, we should also add some of these examples to the test vectors.

File: core/types/rollup_cost_test.go [hunk 5]
```diff
@@ -24,10 +27,12 @@ var (
 	bedrockFee  = big.NewInt(11326000000000)
 	regolithFee = big.NewInt(3710000000000)
 	ecotoneFee  = big.NewInt(960900) // (480/16)*(2*16*1000 + 3*10) == 960900
+	fjordFee    = big.NewInt(63852)  // (-27321890 + 1031462*31 - 88664*30) * (2*16*1000 + 3*10) / 1e6 == 63852
```
Comment (sebastianst) [inline core/types/rollup_cost_test.go]: Sanity check, is it expected that the fee value for the empty tx becomes an order of magnitude smaller? Is this due to using a different scale, e.g. dividing by 16 in the Ecotone cost function? The ratio is pretty close to 16.

File: params/config.go
Comment (sebastianst) [reply params/config.go:1]: And we can already add the Fjord time to `LoadOPStackChainConfig` in `params/superchain.go`

**Review by sebastianst — CHANGES_REQUESTED** [review]

File: core/types/rollup_cost.go [hunk 6]
```diff
@@ -283,3 +336,93 @@ func l1CostHelper(gasWithOverhead, l1BaseFee, scalar *big.Int) *big.Int {
...
+}
+
+func bedrockCalldataGasUsed(costData RollupCostData) (calldataGasUsed *big.Int) {
+	calldataGas := (costData.zeroes * params.TxDataZeroGas) + (costData.ones * params.TxDataNonZeroGasEIP2028)
+	return new(big.Int).SetUint64(calldataGas)
+}
+
+// FlzCompressLen returns the length of the data after compression through FastLZ, based on
+// https://github.com/Vectorized/solady/blob/5315d937d79b335c668896d7533ac603adac5315/js/solady.js
+func FlzCompressLen(ib []byte) uint32 {
```
Comment (roberto-bayardo) [inline core/types/rollup_cost.go:399]: can we have a fuzz test for this?

File: core/types/rollup_cost_test.go (hunk 3 above)
Comment (danyalprout) [reply core/types/rollup_cost_test.go:368]: I'll add this to the monorepo PR for Fjord 👍

File: params/config.go
Comment (danyalprout) [reply params/config.go:1]: Done 👍

File: core/types/rollup_cost.go (hunk 2 above)
Comment (danyalprout) [reply core/types/rollup_cost.go]: Done 👍

File: core/types/rollup_cost.go (hunk 6 above)
Comment (danyalprout) [reply core/types/rollup_cost.go:399]: I'll add this to the monorepo PR w/ the differential test that @sebastianst mentioned above.

File: core/types/rollup_cost_test.go (hunk 3 above)
Comment (danyalprout) [reply core/types/rollup_cost_test.go:368]: @sebastianst quick question on the best way to approach this. Ideally in the monorepo I would like to have a fuzz test that compares the output of the c implementation, the solidity implementation and our op-geth implementation.

I think the solidity comparison requires it to be in `op-e2e` which right now doesn't have any fuzz testing. Any objections to me adding a makefile target/updating circleci as part of this?

Comment (sebastianst) [reply core/types/rollup_cost_test.go:368]: No objections, thanks for adding!

File: core/types/rollup_cost.go (hunk 6 above)
Comment (roberto-bayardo) [reply core/types/rollup_cost.go:399]: Don't feel so strongly about it but since the code is here shouldn't the fuzz test be here too?   We should fuzz test the solidity implementation in the monorepo I think, then differential test would be there too.

Comment (sebastianst) [reply core/types/rollup_cost.go:399]: I think the way Danyal did it now to differential fuzz in the monorepo should be sufficient.

File: core/types/rollup_cost.go (hunk 1 above)
Comment (sebastianst) [reply core/types/rollup_cost.go]: Did we add the rule to the spec that Fjord has to either be activated at genesis, or must activate at least one block after Ecotone, to avoid handling this edge case?

File: core/types/rollup_cost.go [hunk 7]
```diff
@@ -114,6 +123,42 @@ func NewL1CostFunc(config *params.ChainConfig, statedb StateGetter) L1CostFunc {
...
+		l1FeeScalars := statedb.GetState(L1BlockAddr, L1FeeScalarsSlot).Bytes()
+		l1BlobBaseFee := statedb.GetState(L1BlockAddr, L1BlobBaseFeeSlot).Big()
+		l1BaseFee := statedb.GetState(L1BlockAddr, L1BaseFeeSlot).Big()
+
+		if config.IsOptimismEcotone(blockTime) {
+			// Edge case: the very first Ecotone block requires we use the Bedrock cost
+			// function. We detect this scenario by checking if the Ecotone parameters are
+			// unset. Note here we rely on assumption that the scalar parameters are adjacent
+			// in the buffer and l1BaseFeeScalar comes first. We need to check this prior to
+			// other forks, as the first block of Fjord and Ecotone could be the  same block.
```
Comment (sebastianst) [inline core/types/rollup_cost.go]:
```suggestion
			// other forks, as the first block of Fjord and Ecotone could be the same block.
```

File: core/types/rollup_cost.go [hunk 8]
```diff
@@ -114,6 +123,42 @@ func NewL1CostFunc(config *params.ChainConfig, statedb StateGetter) L1CostFunc {
...
+			firstEcotoneBlock := l1BlobBaseFee.BitLen() == 0 &&
+				bytes.Equal(emptyScalars, l1FeeScalars[scalarSectionStart:scalarSectionStart+8])
```
Comment (sebastianst) [inline core/types/rollup_cost.go]: Wondering why we don't do something similar like for op-node and add a `IsOptimismFirstEcotoneBlock(uint64) bool` to the config. In the monorepo we have `IsEcotoneActivationBlock`, which is more straight forward and doesn't check indirect assumptions.

File: core/types/rollup_cost.go [hunk 9]
```diff
@@ -103,6 +111,7 @@ func NewRollupCostData(data []byte) (out RollupCostData) {
 			out.ones++
 		}
 	}
+	out.fastlzSize = uint64(FlzCompressLen(data))
```
Comment (sebastianst) [inline core/types/rollup_cost.go]: I realize that we now _always_ count zeros and non-zeros, _and_ compute the flzCompressLen. Would it be a large lift to modularize the implementation and only calculate the fields that we need at the time the cost function is applied? Probably something to consider for a follow up. I quickly checked all usages, and we could change the `RollupCostData` struct into an interface with getters for the individual fields, and then lazily calculate them on first access.

File: core/types/rollup_cost.go [hunk 10]
```diff
@@ -240,8 +260,30 @@ func extractL1GasParams(config *params.ChainConfig, time uint64, data []byte) (g
 	// edge case: for the very first Ecotone block we still need to use the Bedrock
 	// function. We detect this edge case by seeing if the function selector is the old one
 	// If so, fall through to the pre-ecotone format
+	// Both Ecotone and Fjord use the same function selector
 	if config.IsEcotone(time) && len(data) >= 4 && !bytes.Equal(data[0:4], BedrockL1AttributesSelector) {
```
Comment (sebastianst) [inline core/types/rollup_cost.go:266]: We could also use something like `IsFirstEcotoneBlock` here.

File: core/types/rollup_cost_test.go [hunk 11]
```diff
@@ -237,13 +313,74 @@ func TestNewL1CostFunc(t *testing.T) {
...
+		input       []byte
+		expectedLen uint32
+	}{
+		// empty input
+		{[]byte{}, 0},
+		// all 1 inputs
+		{bytes.Repeat([]byte{1}, 1000), 21},
+		// all 0 inputs
+		{make([]byte, 1000), 21},
+		// empty tx input
```
Comment (sebastianst) [inline core/types/rollup_cost_test.go:397]: Do we have tests in the monorepo that the cost function applied to such input vectors is the min tx length 100? We could also add some of these edge case unit tests for the cost function to this repo instead.

File: core/types/rollup_cost.go (hunk 9 above)
Comment (danyalprout) [reply core/types/rollup_cost.go]: Sounds great -- will do this in a follow up. We have a pairing session this week and it'll be a perfect task 👍

File: core/types/rollup_cost.go (hunk 8 above)
Comment (danyalprout) [reply core/types/rollup_cost.go]: I believe the reason we checked these assumptions when we did the Ecotone upgrade vs. having a similar function, is that we don't have access to the block time param in the op-geth genesis config. (op-node does via the rollup config).

We could add it to the genesis.json file (under config, along w/ block fork times), but it would require modification for all chains running Ecotone.

File: core/types/rollup_cost_test.go (hunk 11 above)
Comment (danyalprout) [reply core/types/rollup_cost_test.go:397]: I added a check for this in op-geth: https://github.com/ethereum-optimism/op-geth/pull/249/files#diff-fb404815e77577216187c4f92206555226fed0d62c39727672e659b0a4965a15R61-R95

To ensure that when the linear regression output is sub 100, the cost func uses 100 as the minimum

File: core/types/rollup_cost_test.go [hunk 12]
```diff
@@ -46,9 +52,64 @@ func TestBedrockL1CostFunc(t *testing.T) {
...
+func TestFjordL1CostFuncMinimumBounds(t *testing.T) {
+	costFunc := NewL1CostFuncFjord(
+		baseFee,
+		blobBaseFee,
+		baseFeeScalar,
+		blobBaseFeeScalar,
+	)
+
+	// Minimum size transactions:
+	// -42.5856 + 0.8365*110 = 49.4294
```
Comment (sebastianst) [inline core/types/rollup_cost_test.go:71]:
```suggestion
	// -42.5856 + 0.8365*100 = 41.0644
```

**Review by sebastianst — APPROVED** [review]
LGTM, thanks!

File: core/types/rollup_cost.go [hunk 13]
```diff
@@ -114,6 +123,42 @@ func NewL1CostFunc(config *params.ChainConfig, statedb StateGetter) L1CostFunc {
 	}
 	forBlock := ^uint64(0)
 	var cachedFunc l1CostFunc
+	selectFunc := func(blockTime uint64) l1CostFunc {
+		// Note: the various state variables below are not initialized from the DB until this
+		// point to allow deposit transactions from the block to be processed first by state
+		// transition.  This behavior is consensus critical!
+		l1FeeScalars := statedb.GetState(L1BlockAddr, L1FeeScalarsSlot).Bytes()
+		l1BlobBaseFee := statedb.GetState(L1BlockAddr, L1BlobBaseFeeSlot).Big()
+		l1BaseFee := statedb.GetState(L1BlockAddr, L1BaseFeeSlot).Big()
```
Comment (protolambda) [inline core/types/rollup_cost.go:136]: Nit: with this change, it reads the ecotone scalars and blob-basefee data, even in bedrock blocks. It doesn't affect consensus, but does add a little overhead.

File: core/types/rollup_cost.go [hunk 14]
```diff
@@ -311,3 +349,125 @@ func l1CostHelper(gasWithOverhead, l1BaseFee, scalar *big.Int) *big.Int {
...
 	return fee
 }
+
+// NewL1CostFuncFjord returns an l1 cost function suitable for the Fjord upgrade
+func NewL1CostFuncFjord(l1BaseFee, l1BlobBaseFee, l1BaseFeeScalar, l1BlobBaseFeeScalar *big.Int) l1CostFunc {
+	return func(costData RollupCostData) (fee, calldataGasUsed *big.Int) {
+		// Fjord L1 cost function:
+		//l1FeeScaled = baseFeeScalar*l1BaseFee*16 + blobFeeScalar*l1BlobBaseFee
+		//estimatedSize = intercept + fastlzCoef*fastlzSize
+		//l1Cost = max(minTransactionSize, estimatedSize) * l1FeeScaled / 1e12
```
Comment (protolambda) [inline core/types/rollup_cost.go]: Specs say:
```
l1FeeScaled = baseFeeScalar*l1BaseFee*16 + blobFeeScalar*l1BlobBaseFee
estimatedSize = max(minTransactionSize, intercept + fastlzCoef*fastlzSize)
l1Cost = estimatedSize * l1FeeScaled / 1e12
```
While I understand that this is effectively the same math, consistency with the intermediate variable naming would be nice.

File: core/types/rollup_cost.go [hunk 15]
```diff
@@ -311,3 +349,125 @@ func l1CostHelper(gasWithOverhead, l1BaseFee, scalar *big.Int) *big.Int {
...
+		calldataCostPerByte := new(big.Int).Mul(l1BaseFee, sixteen)
+		calldataCostPerByte.Mul(calldataCostPerByte, l1BaseFeeScalar)
+		blobCostPerByte := new(big.Int).Mul(l1BlobBaseFee, l1BlobBaseFeeScalar)
+		l1FeeScaled := new(big.Int).Add(calldataCostPerByte, blobCostPerByte)
+
+		l1CostSigned := new(big.Int).Set(fastlzSize)
+		l1CostSigned.Mul(l1CostSigned, l1FeeScaled)
+		l1CostSigned.Div(l1CostSigned, fjordDivisor)
+
+		return l1CostSigned, calldataGasUsed
```
Comment (protolambda) [inline core/types/rollup_cost.go]: nit: The variable naming / reuse is a bit confusing, and the `calldataGasUsed` computation in-between the computation of `l1CostSigned` also threw me off a little while comparing to the specs.

**Review by protolambda — COMMENTED** [review]
Generally looks good. `FlzCompressLen` is difficult to review in isolation, but the tests in the monorepo for it look good. Still looking into the differences with Brotli.

**Review by protolambda — APPROVED** [review]
